File: tinysat-python/solvers/dpll_solver.py
import copy
import random

from enum import Enum

from base.instance import Instance
from base.propagator import Propagator
from base.helper import lit_to_var, var_to_lit


class DPLLSolver:
    """A SAT solver using DPLL algorithm. 

    Parameters: \\
    `var_choice`: {"input_order", "most_appearance"}, optional.
        Default is "input_order".
    `solve_all`: boolean, optional. Default is False, which means the solver \
        will terminate after having found the first solution.
    """

    def __init__(self, var_choice="input_order", solve_all=False):        
        # TODO: Implement the algorithm to solve all possible assignment.

        self.instance = None
        self.assignment = [] # assignment[v]: {True, False, None}
        self.propagator = None
        self.unassigned_vars = []
        self.satisfy_assignment = [] # list of satisfying assignment found.

        self.var_choice = var_choice

    def solve(self, instance):
        self.instance = instance
        self.assignment = [None for _ in range(instance.var_count)]
        self.propagator = Propagator(instance)
        self.unassigned_vars = self.setup_branch_order(instance)
        
        # print(f"\nBranch order: {[x+1 for x in self.unassigned_vars]}")

        self.dpll(0)
        return len(self.get_assignments()) > 0

    def dpll(self, level):
        """
        `level`: the current decision level. Used for backtracking.
        """
        if level == self.instance.var_count:
            self.save_assignment()
            # print(f"Successful assignment:\n {self.assignment}")

            return True
        
        # Pick a variable & try assigning value to it:
        variable = self.pick_branching_var()
        values = self.pick_assignment_order()

        self.assign(variable, values[0])
        if self.propagator.propagate(var_to_lit(variable, values[0]), self.assignment):
            if self.dpll(level + 1):
                return True
        
        self.assign(variable, values[1])
        if self.propagator.propagate(var_to_lit(variable, values[1]), self.assignment):
            if self.dpll(level + 1):
                return True
                
        # Roll back to backtrack:
        self.unassign(variable)
        return False


    def setup_branch_order(self, instance):
        if self.var_choice == "most_appearance":
            all_literals = [l for clause in instance.clauses for l in clause]

            # Sort variables according to their decreasing number of appearances
            order = copy.copy(instance.variables)
            order.sort(key=lambda v: all_literals.count(var_to_lit(v, True))
                                   + all_literals.count(var_to_lit(v, False)),
                reverse=True                                
            )
        else:
            order = copy.copy(instance.variables)

        return order

    def pick_branching_var(self):
        """ Choose the unassigned variable according to self.var_choice"""
        var = self.unassigned_vars[0]
        return var

    def pick_assignment_order(self):
        """ Randomly choose the order of value to make an assignment """
        if random.random() >= 0.5:
            return (True, False)
        return (False, True)

    def assign(self, variable, value):
        self.assignment[variable] = value
        if variable in self.unassigned_vars:
            self.unassigned_vars.remove(variable)

    def unassign(self, variable):
        self.assignment[variable] = None
        self.unassigned_vars.append(variable)        

    def save_assignment(self):
        self.satisfy_assignment.append(copy.copy(self.assignment))

    def get_assignments(self):
        """ Return the list of satisfying assignments.
        """
        return self.satisfy_assignment

File: tinysat-java/src/tinysat/solvers/DPLLSolver.java
package tinysat.solvers;

import tinysat.base.Instance;
import tinysat.base.Propagator;
import tinysat.base.SolverOption;
import tinysat.base.Variable;

import java.util.*;

public class DPLLSolver implements Solver {
    private Instance instance;
    private SolverOption option;
    private List<Variable> assignment;
    private List<List<Variable>> satisfyAssignments;
    private Propagator propagator;

    public DPLLSolver() {
        this(new SolverOption());
    }

    public DPLLSolver(SolverOption option)  {
        this.option = option;
    }

    @Override
    public void solve(Instance instance) {
        this.instance = instance;
        this.assignment = setupBranchOrder(instance);
        this.propagator = new Propagator(instance);
        this.satisfyAssignments = new ArrayList<>();

        dpll(0);
    }

    // Recursive version of dpll
    private boolean dpll(int depth) {
        if (depth == instance.getVarCount()) {
            saveAssignment();
            return true;
        }

        // Pick a variable and try assigning value to it:
        Variable var = assignment.get(depth);
        boolean values[] = pickAssignmentOrder();

        var.assign(values[0]);
        if (propagator.propagate(var.toLit(), this.assignment)) {
            if (dpll(depth + 1)) {
                return true;
            }
        }

        var.assign(values[1]);
        if (propagator.propagate(var.toLit(), this.assignment)) {
            if (dpll(depth + 1)) {
                return true;
            }
        }

        // Could not find solution -> roll back to backtrack:
        var.unassign();
        return false;
    }

    // Iterative version of dpll
    private boolean dpll() {
        int depth = 0;

        // State = -1 means nothing is tried,
        // 0 means False has been tried but True has not,
        // 2 means both value has been tried.
        int[] state = new int[instance.getVarCount()];
        Arrays.fill(state, -1);

        while (true) {
            if (depth == instance.getVarCount()) {
                saveAssignment();
                return true;
            }

            boolean triedSomething = true;
            // Pick a variable and try assigning value to it:
            Variable var = assignment.get(depth);
            switch (state[depth]) {
                case -1:
                    boolean[] values = pickAssignmentOrder();
                    var.assign(values[0]);
                    state[depth] = values[0] ? 1 : 0;
                    break;
                case 0:
                    var.assign(true);
                    state[depth] = 2;
                    break;
                case 1:
                    var.assign(false);
                    state[depth] = 2;
                    break;
                case 2:
                    // Backtrack
                    triedSomething = false;
                    break;
            }
            if (triedSomething) {
                if (propagator.propagate(var.toLit(), this.assignment)) {
                    ++depth;
                } else {
                    var.unassign();
                }
            } else {
                // Backtrack:
                if (depth > 0) {
                    var.unassign();
                    state[depth] = -1;
                    --depth;
                } else {
                    return false;
                }
            }
        }
    }

    private void saveAssignment() {
        List<Variable> assignment = new ArrayList<>();
        for (Variable v : this.assignment) {
            assignment.add(v.clone());
        }
        this.satisfyAssignments.add(assignment);
    }

    @Override
    public List<List<Variable>> getAssignments() {
        return this.satisfyAssignments;
    }

    private List<Variable> setupBranchOrder(Instance instance) {
        List<Variable> branchOrder = new ArrayList<>();
        for (int id = 1; id <= instance.getVarCount(); ++id) {
            branchOrder.add(new Variable(id));
        }

        if (this.option.getVarChoice() == SolverOption.VarChoice.INPUT_ORDER) {
            // Do nothing
        } else {
            // TODO: Change to MostAppearance using statistics
//            branchOrder.addAll(listVars);
        }

        return branchOrder;
    }

    private boolean[] pickAssignmentOrder() {
        Random rnd = new Random();
        boolean value = rnd.nextBoolean();
        return new boolean[] {value, !value};
//        return new boolean[] {true, false};
    }
}
